tests/tck-build-logic/src/main/java/org/graalvm/internal/tck/ContributionTask.java

```diff
@@ -0,0 +1,586 @@
+        int replaceEntryIndex = -1;
+        for (int i = 0; i < entries.size(); i++) {
+            if (entries.get(i).module().equals(coordinates.group() + ":" + coordinates.artifact())) {
+                replaceEntryIndex = i;
+            }
+        }
+
+        if (replaceEntryIndex != -1) {
+            MetadataIndexEntry replacedEntry = entries.remove(replaceEntryIndex);
+            Set<String> extendedAllowedPackages = new HashSet<>(replacedEntry.allowedPackages());
+            extendedAllowedPackages.addAll(allowedPackages);
+
+            entries.add(new MetadataIndexEntry(replacedEntry.directory(), replacedEntry.module(), replacedEntry.requires(), new ArrayList<>(extendedAllowedPackages)));
+        }
+
+        List<MetadataIndexEntry> sortedEntries = entries.stream()
+                .sorted(Comparator.comparing(MetadataIndexEntry::module))
+                .toList();
+
+        objectMapper.writeValue(metadataIndex, sortedEntries);
+    }
+
+    private void addTests(Path originalTestsLocation){
+        Path destination = testsDirectory.resolve("src")
+                .resolve("test")
+                .resolve("java");
+        Path allTests = originalTestsLocation.resolve(".");
+
+        InteractiveTaskUtils.printUserInfo("Removing dummy test stubs");
+        invokeCommand("rm -r " + destination, "Cannot delete files from: " + destination);
+
+        InteractiveTaskUtils.printUserInfo("Copying tests from: " + originalTestsLocation + " to " + destination);
+        invokeCommand("cp -a " + allTests + " " + destination, "Cannot copy files to: " + destination);
+    }
+
+    private void addResources(Path originalResourcesDirectory){
+        if (originalResourcesDirectory == null) {
+            return;
+        }
+
+        Path destination = testsDirectory.resolve("src").resolve("test");
+        InteractiveTaskUtils.printUserInfo("Copying resources from: " + originalResourcesDirectory + " to " + destination);
+
+        invokeCommand("cp -r " + originalResourcesDirectory + " " + destination, "Cannot copy files to: " + destination);
+    }
+
+    private void addDockerImages(List<String> images) throws IOException {
+        if (images.isEmpty()) {
+            return;
+        }
+
+        InteractiveTaskUtils.printUserInfo("Adding following docker images to " + REQUIRED_DOCKER_IMAGES_FILE + ": " + images);
+        Path destination = testsDirectory.resolve(REQUIRED_DOCKER_IMAGES_FILE);
+        if (!Files.exists(destination)) {
+            Files.createFile(destination);
+        }
+
+        for (var image : images) {
+            writeToFile(destination, image.concat("\n"), StandardOpenOption.APPEND);
+        }
+    }
+
+    private void addUserCodeFilterFile(List<String> packages) throws IOException {
+        InteractiveTaskUtils.printUserInfo("Generating " + USER_CODE_FILTER_FILE);
+
+        ConfigurationStringBuilder sb = new ConfigurationStringBuilder();
```

Comment:
Wouldn't be an object which is then written to JSON via the `ObjectMapper` better here?